workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (badTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public IEnumerable<uint> GetRequiredReaderComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentReaderIdsForBehaviours[behaviourType];
+        }
+
+        public IEnumerable<uint> GetRequiredWriterComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentWriterIdsForBehaviours[behaviourType];
+        }
+
+        private void Inject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var readerWriter = ReaderWriterFactory.CreateReaderWriter(componentId);
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, readerWriter);
+        }
+
+        private void DeInject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, null);
+        }
+
+        private void EnsureLoaded(Type behaviourType)
+        {
+            if (adapterCache.ContainsKey(behaviourType))
+            {
+                return;
+            }
+
+            var adapters = reflectionUtil.GetMembersWithMatchingAttributes(behaviourType);
+            var componentIdsToAdapters = new Dictionary<uint, IMemberAdapter>();
+            var readerIds = new List<uint>();
+            var writerIds = new List<uint>();
+            adapterCache[behaviourType] = componentIdsToAdapters;
+            componentReaderIdsForBehaviours[behaviourType] = readerIds;
+            componentWriterIdsForBehaviours[behaviourType] = writerIds;
+            foreach (var adapter in adapters)
+            {
+                // Figure out if reader or writer
+                // Get component ID
+                // Store in data structures
+                Type requiredType = adapter.TypeOfMember;
+                var componentIdAttribute =
+                    (ComponentIdAttribute) Attribute.GetCustomAttribute(requiredType, typeof(ComponentIdAttribute), false);
+                if (componentIdAttribute == null)
+                {
+                    logger.HandleLog(LogType.Warning, new LogEvent(BadRequiredMemberWarning)
```

Comment:
This warning doesn't exactly correspond to the condition under which it is fired.

This will fire if the underlying type injected doesn't have a component id attribute - this doesn't tell you whether the underlying type is a reader or writer at all.

Reply:
The idea is that all reader and writer interfaces have to be generated with the component ID attribute as well, so if something doesn't have it, it must be something else or a malformed reader/writer or something.

Reply:
I think the problem is this error is used in two places for two actually different errors:

1. A Reader/Writer doesn't have a `[ComponentId]` attribute
2. The thing marked as `[Require]` is not a Reader/Writer

Reply:
This will convolute the control flow slightly but I'll make it so that there's a separate error message if there is no Reader/Writer marks on the required thing, or if there's Reader/Writer but no ComponentId. The latter really shouldn't happen though.